File: 2019-2020/December/Silver/SwapitySwapitySwap.java
import java.util.*;
import java.io.*;

public class SwapitySwapitySwap {
  static int n, m, k;
  static int[] arr;
  static int[] mark;
  static int counter = 1;
  public static void main(String[] args) throws IOException {
    BufferedReader br = new BufferedReader(new FileReader("swap.in"));
    PrintWriter pw = new PrintWriter(new FileWriter("swap.out"));
    StringTokenizer st = new StringTokenizer(br.readLine());
    n = Integer.parseInt(st.nextToken());
    m = Integer.parseInt(st.nextToken());
    k = Integer.parseInt(st.nextToken());
    arr = new int[n];
    mark = new int[n];
    for (int i = 0; i < n; i++) arr[i] = i;
    for (int i = 0; i < m; i++) {
      st = new StringTokenizer(br.readLine());
      int a = Integer.parseInt(st.nextToken()) -1;
      int b = Integer.parseInt(st.nextToken()) -1;
      flip(a, b);
    }
    System.out.println(Arrays.toString(arr));
    for (int i = 0; i < n; i++) {
      if (mark[i] == 0) {
        dfs(i);
        counter++;
      }
    }
    HashMap<Integer, Integer> map = new HashMap<Integer, Integer>();
    for (int ele: mark) {
      if (!map.keySet().contains(ele)) map.put(ele, 0);
      map.put(ele, map.get(ele)+ 1);
    }
    for (int i = 0; i < n; i++) {
      if (map.get(mark[i]) == 1) {
        System.out.println(i+1);
        pw.println(i+1);
        continue;
      }
      int rep = k % map.get(mark[i]);
      int bef = i;
      for (int j = 0; j < rep; j++) {
        bef = arr[bef];
      }
      System.out.println(bef+1);
      pw.println(bef+1);
    }
    pw.close();

  }
  static void dfs(int node) {
    mark[node] = counter;
    if (mark[arr[node]] == 0) dfs(arr[node]);
  }
  static void flip(int one, int two) {
    for (int i = one; i <= (one+two)/2; i++) {
      int temp = arr[i];
      arr[i] = arr[one+two-i];
      arr[one+two-i] = temp;
    }
  }
}

File: 2019-2020/December/Silver/SwapitySwapitySwap.cpp
#include <bits/stdc++.h>

using namespace std;
int n, m, k, a, b, counter;
void dfs(int node, int mark[], int arr[]) {
  mark[node] = counter;
  if (mark[arr[node]] == 0) {
    dfs(arr[node], mark, arr);
  }
}
int main() {
  // freopen("input.txt", "r", stdin);
  freopen("swap.in", "r", stdin);
  freopen("swap.out", "w", stdout);
  cin >> n >> m >> k;
  counter = 1;
  int arr[n];
  int mark[n];
  for (int i = 0; i < n; i++) {
    arr[i] = i;
    mark[i] = 0;
  }
  for (int i = 0; i < m; i++) {
    cin >> a>> b;
    a--;b--;
    for (int i = a; i <= (a+b)/2; i++) {
      int temp = arr[i];
      arr[i] = arr[a+b-i];
      arr[a+b-i] = temp;
    }
  }
  for (int i = 0; i < n; i++) {
    if (mark[i] == 0) {
      dfs(i, mark, arr);
      counter++;
    }
  }
  unordered_map<int, int> map;
  for (int ele: mark) {
    map[ele] = map[ele] + 1;
  }
  for (int i = 0; i < n; i++) {
    if (map[mark[i]] == 1) {
      cout << i+1 << endl;
    }
    else {
      int rep = k % map[mark[i]];
      int bef = i;
      for (int j = 0; j < rep; j++) {
        bef = arr[bef];
      }
      cout << bef+1 << endl;
    }
  }
}
